src/crimson/os/seastore/object_data_handler.h

```diff
@@ -30,21 +30,63 @@ struct block_delta_t {
   }
 };
 
+struct overwrite_buf_t : public boost::intrusive_ref_counter<
+    overwrite_buf_t,
+    boost::thread_unsafe_counter> {
+  size_t add(block_delta_t &b) {
+    changes.push_back(b);
+    return changes.size();
+  }
+  bool empty() {
+    return changes.empty();
+  }
+  size_t get_size() const {
+    return changes.size();
+  }
+  void clear() {
+    if (changes.size()) {
+      changes.clear();
+    }
+  }
+  void copy_changes(bufferptr &b, int to) {
+    int idx = 0;
+    for (auto p : changes) {
+      auto iter = p.bl.cbegin();
+      iter.copy(p.bl.length(), b.c_str() + p.offset);
+      idx++;
+      if (idx >= to) {
+	break;
+      }
+    }
+  }
+  std::vector<block_delta_t> changes = {};
+};
+using overwrite_buf_ref = boost::intrusive_ptr<overwrite_buf_t>;
+
 struct ObjectDataBlock : crimson::os::seastore::LogicalCachedExtent {
   using Ref = TCachedExtentRef<ObjectDataBlock>;
 
   std::vector<block_delta_t> delta = {};
 
   interval_set<extent_len_t> modified_region;
+  // to provide the local modified view during transaction
+  overwrite_buf_ref overwrites = nullptr; 
+  int overwrites_idx = 0;
 
   explicit ObjectDataBlock(ceph::bufferptr &&ptr)
     : LogicalCachedExtent(std::move(ptr)) {}
   explicit ObjectDataBlock(const ObjectDataBlock &other)
     : LogicalCachedExtent(other), modified_region(other.modified_region) {}
+  explicit ObjectDataBlock(const ObjectDataBlock &other, share_buffer_t s)
+    : LogicalCachedExtent(other, s), modified_region(other.modified_region),
+      overwrites(other.overwrites), overwrites_idx(other.overwrites_idx) {}
```

Comment:
See my comment above, why need to consider overwrites and overwrites_idx upon changing to mutation-pending ? IIUC they should be invalid before mutation.